keras/src/callbacks/csv_logger.py

```diff
@@ -65,22 +72,20 @@ def handle_value(k):
 
         if self.keys is None:
             self.keys = sorted(logs.keys())
-            # When validation_freq > 1, `val_` keys are not in first epoch logs
-            # Add the `val_` keys so that its part of the fieldnames of writer.
-            val_keys_found = False
-            for key in self.keys:
-                if key.startswith("val_"):
-                    val_keys_found = True
-                    break
-            if not val_keys_found:
-                self.keys.extend(["val_" + k for k in self.keys])
+
+            # Check if the model is expected to produce validation metrics.
+            if hasattr(self.model, "metrics_names"):
```

Comment:
nitpick: you don't need `if hasattr(self.model, "metrics_names"):`, it's a property on the `Model` class:

https://github.com/keras-team/keras/blob/master/keras/src/trainers/trainer.py#L271